6928542: Chinese characters in RTF are not decoded

takiguc opened the pull request:
"character set of font" (font charset) table was created by "Rich Text Format Specification 1.9.1"
https://interoperability.blob.core.windows.net/files/Archive_References/[MSFT-RTF].pdf
It refers windgi.h
https://learn.microsoft.com/en-us/windows/win32/api/wingdi/ns-wingdi-textmetrica

Test files and testcase are in bugid [JDK-6928542](https://bugs.openjdk.org/browse/JDK-6928542)

Additional change:
Special character `\line` should `\n`

Additional information:

Add 2 hash tables
- fcharsetToCP: Predefined conversion table, `fcharset` with number control word, from control word to Java charset name, `fcharset0` refers `windows-1252` Java charset name
- fcharsetTable: Conversion table for each RTF file, `f` control word with number, from integer font numbers to Charset font charsets, In case of `{\f0\fnil\fcharset0 Segoe UI;}`, `0` refers Java Charset `windows-1252`

When RTF Character Set control word (like `\mac`) is used, unmappable character returns \u0000 and it's not written into RTF text..
When fcharset control word is used, unmappable character returns \uFFFD (it's the same as replacement character on decoder), \u0000 is used for DBCS lead byte detection.
If `f` or `par` control word is there and lead byte is remains on byte buffer for decoder, this byte data is as invalid character and write \uFFFD into RTF text.

If `f` control word is used without `fcharset`, `translationTable` char array is used.
If `f` control word is used with `fcharset`, predefined Java Charset name is used (if missing, ISO8859_1 is used for fallback).

**Note:** Following GitHub actions were failed
linux-cross-compile / build (riscv64), I opened following JBS.
> [JDK-8314624](https://bugs.openjdk.org/browse/JDK-8314624) GHA: RISC-V cross-build was failed

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-6928542](https://bugs.openjdk.org/browse/JDK-6928542): Chinese characters in RTF are not decoded (**Bug** - P3)


### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [2f9c45cb](https://git.openjdk.org/jdk/pull/13553/files/2f9c45cbc016ca48642976b5c537a9e3777c6342)
 * [Prasanta Sadhukhan](https://openjdk.org/census#psadhukhan) (@prsadhuk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13553/head:pull/13553` \
`$ git checkout pull/13553`

Update a local copy of the PR: \
`$ git checkout pull/13553` \
`$ git pull https://git.openjdk.org/jdk.git pull/13553/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13553`

View PR using the GUI difftool: \
`$ git pr show -t 13553`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13553.diff">https://git.openjdk.org/jdk/pull/13553.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13553#issuecomment-1515753471)

takiguc commented:
Gentle reminder, could you give your suggestions ?
Thanks.

takiguc commented:
Gentle reminder, could you give reviewer suggestions ?
Thanks.

takiguc commented:
Gentle reminder, can I have a reviewer's comment ?
Thanks.

takiguc commented:
I'd like to start a discussion about Swing's RTF implementation to support font charset.
If you are interested in RTF, please put your comment into this PR or [JDK-6928542](https://bugs.openjdk.org/browse/JDK-6928542).
Non-reviewers are also welcome.

prrace commented on src/java.desktop/share/classes/javax/swing/text/rtf/RTFParser.java:
```diff
@@ -360,4 +371,34 @@ public void close()
 
+    private char[] ca = new char[1];
+    private CharBuffer decoderCB = CharBuffer.wrap(ca);
+
```
So I know in theory what needs to be done here, and the test looks thorough,
but basically you've dumped code on us with zero explanation of the fix,
and I've spent quite a while figuring out some of it but I think you owe it to reviewers to do a much better job
of explaining the logic and flow. 
The function below needs particular explanation. 
One thing (not the only thing !) I want you to describe is what you are using \uFFFD for in various places in this change.

prrace commented on src/java.desktop/share/classes/javax/swing/text/rtf/RTFReader.java:
```diff
@@ -161,6 +171,48 @@ class RTFReader extends RTFParser
 
 /* TODO: per-font font encodings ( \fcharset control word ) ? */
 
```
Umm, since you are doing it, you really should also delete the TODO comment

takiguc commented:
Sorry, it seems I did wrong merge operation...

takiguc commented:
Hello @prrace .
I'm very sorry for your confusion and I appreciate your comments.
I put some comments into source files and added/updated some lines, and updated description text.
Please let me know if it's not enough.

prrace commented on src/java.desktop/share/classes/javax/swing/text/rtf/RTFParser.java:
```diff
@@ -382,22 +392,24 @@ private char decode(char ch) {
-                return c;
+                return ca[0];
             } else {
+                // Detectied lead byte
```
Minor: detectied -> detected

takiguc commented:
Hello @prsadhuk .
Sorry for bad reaction.
I tried to run my testcase on Japanese Windows 10/11.
It worked fine as expected.
Could you show me your windows environment ?

prsadhuk reviewed (approved):
It was a mistake from my part..It seems to work ok..
LGTM..

takiguc commented:
@prsadhuk , thanks.

/integrate
